trigger.tex

```diff
@@ -39,9 +40,85 @@ \chapter{Trigger Module}
     bits of \RcsrTselect but fewer than $2^m$ triggers.
 \end{commentary}
 
-It is possible for a trigger with the ``enter Debug Mode'' action (1) and another
-trigger with the ``raise a breakpoint exception'' action (0) to fire at the same
-time. The preferred behavior is to have both actions take place.  It is
+\section{Actions}
+
+Triggers can be configured to take one of several actions when they fire.
+Table~\ref{tab:action} lists all options.
+
+\begin{table}[H]
+\centering
+\caption{\FcsrMcontrolAction encoding}
+\label{tab:action}
+\begin{tabular}{|r|L|}
+\hline
+Value & Description \\
+\hline
+0 & Raise a breakpoint exception into M-Mode. (Used when software wants to
+    use the trigger module without an external debugger attached.)
+    If the trigger is precise, \Rmepc must contain the address of the
+    instruction that caused the trigger to fire. If the trigger is imprecise
+    \Rmepc will contain the address of the next instruction to be executed. \\
+\hline
+1 & Enter Debug Mode. (Only supported when the trigger's \FcsrTdataOneDmode
+    is 1.)
+    If the trigger is precise, \RcsrDpc must contain the address of the
+    instruction that caused the trigger to fire. If the trigger is imprecise
+    \RcsrDpc will contain the address of the next instruction to be executed. \\
+\hline
+2 -- 5 & Reserved for use by the trace specification. \\
+\hline
+other & Reserved for future use. \\
+\hline
+\end{tabular}
+\end{table}
+
+\section{Priority}
+
+Table~\ref{tab:priority} lists the synchronous exceptions from the Privileged
+Spec, and where the various types of triggers fit in. The first 3 columns come
+from the Privileged Spec, and the final column shows where triggers fit in. If
+this table contradicts the table in the Privileged Spec, then the latter takes
+precedence. This table only applies if triggers are precise. Otherwise triggers
+will fire some indeterminate time after the event, and the priority is
+irrelevant.
+
+\begin{table}[H]
+\centering
+\label{tab:priority}
+\begin{tabular}{|l|r|l|l|}
+  \hline
+  Priority      & Exception & Description & Trigger \\
+                &      Code &             & \\
+  \hline
+  {\em Highest} &          3 & & icount \\ \hline
+                &          3 & Instruction address breakpoint & mcontrol address before execute \\ \hline
+                &         12 & Instruction page fault & \\ \hline
+                &          1 & Instruction access fault & \\ \hline
+                &          3 & & mcontrol data before execute \\ \hline
```

Comment:
In looking at table 5.8, I'm reminded that there are execute address+instruction breakpoints via chain=1.

I think that there needs to be a sentence that when triggers are chained, the priority is the lowest priority of the triggers in the chain.  That means that execute address+data will be treated like execute data regardless of the order of the address and data triggers in the chain.

On a related note: Should there be a requirement that higher priority things come earlier in the chain?  If trigger 0 is execute data chain=1 and trigger 1 is execute address chain=0, it seems weird to say that the execute data trigger (which comes after fetching the instruction) is preventing the execute address trigger (which comes before fetching the instruction) from matching.  It would be more natural for something earlier in the process to prevent something later in the process from matching.

Reply:
Extra sentence makes sense.

I don't see a big advantage to forcing the debugger to set up the chain in a specific order. The way I imagine it, at each "stage" of executing an instruction the hardware checks any trigger conditions, and sets a hit bit (possibly internal) for each matching trigger. Then it looks at those hit bits and the chain bits and causes a trigger when all that looks OK. The order is irrelevant then.